pkg/certgenerator/v1beta1/generator.go

```diff
@@ -0,0 +1,215 @@
+	configv1beta1 "github.com/kubeflow/katib/pkg/apis/config/v1beta1"
+	"github.com/kubeflow/katib/pkg/controller.v1beta1/consts"
+)
+
+var (
+	errServiceNotFound      = errors.New("unable to locate controller service")
+	errCertCheckFail        = errors.New("failed to check if certs already exist")
+	errCreateCertFail       = errors.New("failed to create certs")
+	errCreateCertSecretFail = errors.New("failed to create secret embedded certs")
+	errInjectCertError      = errors.New("failed to inject certs into WebhookConfigurations")
+)
+
+// CertGenerator is the manager to generate certs.
+type CertGenerator struct {
+	namespace   string
+	serviceName string
+	secretName  string
+	kubeClient  client.Client
+	certsReady  chan struct{}
+
+	certs             *certificates
+	fullServiceDomain string
+}
+
+var _ manager.Runnable = &CertGenerator{}
+var _ manager.LeaderElectionRunnable = &CertGenerator{}
+
+func (c *CertGenerator) Start(ctx context.Context) error {
+	if err := c.generate(ctx); err != nil {
+		return err
+	}
+	// Close a certsReady means start to register controllers to the manager.
+	close(c.certsReady)
+	return nil
+}
+
+func (c *CertGenerator) NeedLeaderElection() bool {
+	return true
+}
+
+// AddToManager adds the cert-generator to the manager.
+func AddToManager(mgr manager.Manager, config configv1beta1.CertGeneratorConfig, certsReady chan struct{}) error {
+	return mgr.Add(&CertGenerator{
+		namespace:   consts.DefaultKatibNamespace,
+		serviceName: config.WebhookServiceName,
+		secretName:  config.WebhookSecretName,
+		kubeClient:  mgr.GetClient(),
+		certsReady:  certsReady,
+	})
+}
+
+// generate generates certificates for the admission webhooks.
+func (c *CertGenerator) generate(ctx context.Context) error {
+	controllerService := &corev1.Service{}
+	if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: c.serviceName, Namespace: c.namespace}, controllerService); err != nil {
+		return fmt.Errorf("%w: %v", errServiceNotFound, err)
+	}
+
+	certExist, err := c.isCertExist(ctx)
+	if err != nil {
+		return fmt.Errorf("%w: %v", errCertCheckFail, err)
+	}
+	if !certExist {
+		c.fullServiceDomain = strings.Join([]string{c.serviceName, c.namespace, "svc"}, ".")
+
+		if err = c.createCert(); err != nil {
+			return fmt.Errorf("%w: %v", errCreateCertFail, err)
+		}
+		if err = c.updateCertSecret(ctx); err != nil {
+			return fmt.Errorf("%w: %v", errCreateCertSecretFail, err)
+		}
+	}
+	if err = c.injectCert(ctx); err != nil {
+		return fmt.Errorf("%w: %v", errInjectCertError, err)
+	}
+	return nil
+}
+
+// isCertExist checks if a secret embedded certs already exists.
+// For example, it will return true if the katib-controller is created with enabled leader-election
+// since another controller pod will create the secret.
+func (c *CertGenerator) isCertExist(ctx context.Context) (bool, error) {
+	secret := &corev1.Secret{}
+	if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: c.secretName, Namespace: c.namespace}, secret); err != nil {
```

Comment:
I think, if we are going with `secret` approach, we need to create this secret as part of cert-generator, but in that case how [certs from secret can be fetched](https://github.com/kubeflow/katib/blob/eac65bf8e31fe4d3a2c4dbb20be30d600c086157/manifests/v1beta1/components/controller/controller.yaml#L61-L65) if Deployment is already started ?

Reply:
> we need to create this secret as part of cert-generator, but in that case how [certs from secret can be fetched](https://github.com/kubeflow/katib/blob/eac65bf8e31fe4d3a2c4dbb20be30d600c086157/manifests/v1beta1/components/controller/controller.yaml#L61-L65) if Deployment is already started ?

Yes, there is a consideration of whether the controller starts first or the certs are generated last.

So I created an empty secret and then the controller mounts the secret:

https://github.com/kubeflow/katib/blob/eac65bf8e31fe4d3a2c4dbb20be30d600c086157/manifests/v1beta1/components/webhook/secret.yaml

Reply:
@tenzen-y Will controller fetch the updated data from secret[ after we do patch the secret](https://github.com/kubeflow/katib/blob/eac65bf8e31fe4d3a2c4dbb20be30d600c086157/pkg/certgenerator/v1beta1/generator.go#L180) ?

Reply:
@andreyvelich Yes, that's right. Please refer to the kubernetes doc: https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-files-from-a-pod

Reply:
Makes sense, thanks for sharing!